### PR TITLE
Implement a request fingerprinter that accounts for dependencies

### PR DESCRIPTION
**Pull request by Gallaecio**
Pending decisions:
- [x] [`DummyResponse`](https://github.com/scrapinghub/scrapy-poet/pull/172#discussion_r1391433720) (ignore)
- [x] [Ignoring specific page inputs](https://github.com/scrapinghub/scrapy-poet/pull/172#discussion_r1391450915) (yes, ignore them)
- [x] [Dependency resolution](https://github.com/scrapinghub/scrapy-poet/pull/172#issuecomment-1808641236) (apply now that `ItemProvider` is gone)
- [x] [Page params](https://github.com/scrapinghub/scrapy-poet/pull/172#issuecomment-1808726190) (expect JSON-serializable page params)

### CODE REVIEW

File: scrapy_poet/_request_fingerprinter.py
Comment (Gallaecio) [inline scrapy_poet/_request_fingerprinter.py:79]: We could technically save this call to sha1, and instead let callers use the generated bytes to generate their own hash, which they are expected to do anyway.

But this feels cleaner for some reason, and with callback caching, it should be rather cheap.

File: tests/test_request_fingerprinter.py
Comment (BurnzZ) [inline tests/test_request_fingerprinter.py:54]: The spider is the same here with the one above. We can probably remove this and instead add a 3rd test above:

```python
    request3 = Request("https://toscrape.com", callback=crawler.spider.parse_page)
    fingerprint3 = fingerprinter.fingerprint(request3)
    assert fingerprint2 == fingerprint2
```

Also adding a comment to emphasize better that its testing for same deps.

File: tests/test_request_fingerprinter.py
Comment (BurnzZ) [inline tests/test_request_fingerprinter.py:37]: Let's also add a test with different URLs to get some smoke when refactoring in the future, just in case.

Comment (BurnzZ) [reply tests/test_request_fingerprinter.py:37]: Let's also add one with `DummyResponse` as well.

File: scrapy_poet/_request_fingerprinter.py
Comment (BurnzZ) [inline scrapy_poet/_request_fingerprinter.py:51]: Unrelated note: I find myself needed to access this as well when debugging. Would it be worth it having this feature in Scrapy itself? Perhaps some getter method in `Crawler` to easily grab some middleware.

**Review by BurnzZ — COMMENTED** [review]
I was thinking earlier today if there are other approaches aside from this but it seems the Request Fingerprinter route is the best way to go 👍

File: tests/__init__.py
Comment (Gallaecio) [reply tests/__init__.py:19]: I actually copied it from scrapy-zyte-api, so there is definitely room for reuse.

File: tests/__init__.py
Comment (kmike) [inline tests/__init__.py:44]: What do you think about moving this code to e.g. tests.utils, to keep the `__init__` minimal?

File: scrapy_poet/_request_fingerprinter.py
Comment (kmike) [inline scrapy_poet/_request_fingerprinter.py:65]: I wonder if using fq path can be better than repr. `repr(cls)` includes the id, right? So, does it mean it's not going to be stable across runs? I think we need fingerprints to be stable, e.g. for cache to work.

File: tests/test_request_fingerprinter.py
Comment (Gallaecio) [reply tests/test_request_fingerprinter.py:37]: Hmm… Regarding `DummyResponse`, I 100% agree on having the test, but I wonder what the outcome should be. Right now the typing of the response is ignored, and I think that can be fine, but it could also be argued that `DummyResponse` should generate a different fingerprint (since in some cases that could result in different underlying requests).

**Comment by Gallaecio** [comment]
I’m adding a test, `test_dep_resolution`, that shows how fingerprinting does *not* resolve dependencies, so callbacks that effectively request the same dependencies could end up with a different fingerprint.

I assume this is OK given how messy dependency resolution is due to ItemProvider, and how unlikely such cases are (having 2 callbacks on the same spider that request the same deps indirectly). However, if we ever [get rid of ItemProvider](https://github.com/scrapinghub/scrapy-poet/pull/151#discussion_r1386931164), it might be worth considering.

File: tests/test_request_fingerprinter.py
Comment (Gallaecio) [reply tests/test_request_fingerprinter.py:37]: On a related note, I wonder if we should ignore request inputs (e.g. RequestUrl, RequestBody, RequestHeaders) for fingerprinting, given they cannot alter the result.

**Comment by Gallaecio** [comment]
One more thing: page params.

I have just realized that we should probably alter the fingerprint based on their presence in `Request.meta`, but [web-poet says it is an arbitrary dict](https://web-poet.readthedocs.io/en/stable/api-reference.html#web_poet.page_inputs.page_params.PageParams), so it might not be serializable and thus fingerprintable.

- Should we try to take page params into account any way for request fingerprinting?
- Should we log a warning if it is not serializable?
- How should we fingerprint requests with unserializable parameters?
    - Same fingerprint as requests without parameters
    - Their own fingerprint, but common to all requests with unserializable parameters (fixed salt). i.e. identical requests with different unserializable parameters are considered the same nonetheless.
    - Their own fingerprint, unique for each unserializable parameter set (random salt). i.e. identical requests with the same unserializable parameters are considered different nonetheless.

File: scrapy_poet/_request_fingerprinter.py
Comment (Gallaecio) [inline scrapy_poet/_request_fingerprinter.py:31]: @kmike Is this `repr` OK?

File: tests/test_request_fingerprinter.py
Comment (BurnzZ) [reply tests/test_request_fingerprinter.py:38]: >  but I wonder what the outcome should be. Right now the typing of the response is ignored, and I think that can be fine

+1 on this. 

> but it could also be argued that DummyResponse should generate a different fingerprint (since in some cases that could result in different underlying requests).

The resulting different fp would pretty much be based on the other injected parameters anyways, so it's okay to be ignored.

>  I wonder if we should ignore request inputs (e.g. RequestUrl, RequestBody, RequestHeaders) for fingerprinting, given they cannot alter the result.

Hmm not sure about this, as there might be changes in the future that affects them. For example, `typing.Annotated` might be introduced to them.

File: tests/test_request_fingerprinter.py
Comment (BurnzZ) [inline tests/test_request_fingerprinter.py:58]: I think let's also add a test case for `Request.meta` where it doesn't affect the fps as well.

Although I'm not too sure about this since if the `"page_params"` key is present, it would create `PageParams` which _might_ affect how a Page Object behaves. For example, a boolean flag is passed to run additional requests.

This is a complicated example and might be too specific. But having the fp equality tests for now would be useful later on when we decide to do something about it.

File: tests/test_request_fingerprinter.py
Comment (Gallaecio) [reply tests/test_request_fingerprinter.py:38]: > The resulting different fp would pretty much be based on the other injected parameters anyways, so it's okay to be ignored.

I‘m thinking of:

```
def parse(self, response, request_url: RequestUrl)
def parse(self, response: DummyResponse, request_url: RequestUrl)
```

It is an unrealistic scenario, though, given that the use of `DummyResponse` is meant for cases where you get a response differently (e.g. BrowserHtml), scenarios where you don’t get a response at all seem impractical, so yes, I think you are right.

Comment (Gallaecio) [reply tests/test_request_fingerprinter.py:70]: > Hmm not sure about this, as there might be changes in the future that affects them. For example, typing.Annotated might be introduced to them.

What about [ignoring unannotated request and utility page inputs](https://github.com/scrapinghub/scrapy-poet/pull/172/commits/a611fd2714f60135af8299d184697dda1f9c3ac0) for now, and deciding what to do with their annotated versions later on (e.g. after merging https://github.com/scrapinghub/scrapy-poet/pull/169)?

File: tests/test_request_fingerprinter.py
Comment (Gallaecio) [inline tests/test_request_fingerprinter.py:315]: I think we can agree that resolving dependencies now would be too hard given ItemProvider. But I think the question then becomes: do we keep this PR on hold until we can easily resolve dependencies, or do we merge without dependency resolving and then later on change all fingerprints (invalidating caches)?

File: scrapy_poet/_request_fingerprinter.py
Comment (Gallaecio) [inline scrapy_poet/_request_fingerprinter.py:154]: Given [we expect JSON-serializable page params for caching](https://github.com/scrapinghub/web-poet/blob/d45aa1e74aab15466100cbf71b1e655d6c6572f0/web_poet/serialization/functions.py#L175), I went with expecting so, and logging a warning and fingerprinting as if there were no params when that’s not the case.

I thought about keeping a fixed size cache for page params, so save json.dumps calls on scenarios where the same page params are often used, but I was unsure about it.

I also thought about using a random value instead of treating them the same as missing or empty params, to force a different fingerprint, but that could cause a different fingerprint for requests with the same (unserializable) params, which I think is worse (can destroy more data through duplicate request dropping).

File: scrapy_poet/_request_fingerprinter.py
Comment (kmike) [inline scrapy_poet/_request_fingerprinter.py:44]: Should we use https://github.com/scrapinghub/web-poet/blob/acef0a6a6d6dc8598176fc401d5785d1c9c4bf6a/web_poet/utils.py#L25?

File: scrapy_poet/_request_fingerprinter.py
Comment (kmike) [reply scrapy_poet/_request_fingerprinter.py:43]: It should work fine with the current features, but there are caveats to keep in mind - the repr of annotated values shouldn't contain ids. Python docs mention addresses commonly being a part of repr, see https://docs.python.org/3/library/functions.html#repr. For example, if we define an annotation like this (likely we won't be adding geolocation support this way though), it'd break:

```py
class Geolocation:
    def __init__(self, geo: str):
        self.geo = geo

@attrs.define
class MyPage(ItemPage):
    resp: Annotated[HttpResponse, [Geolocation("UK")]]
```

To fix it, developer should override `__repr__` of the Geolocation class; it looks non-obvious than `__repr__` changes the fingerprinting behavior. 

I'm not sure what's the best fix here. One option is to require annotations to be json-serializable, but that's invasive, as even Enum values are not serializable. 

We can also solve it with a piece of documentation, to mention that repr is used for fingerprints. 

It may also be an option to add support for some other special method to be used to serialize annotations, which the classes can define, and fallback to repr if it's not available. It may also be a singledispatch hook, not a method.

I'd probably go with docs approach for now, because the issue doesn't seem to affect any of the current or planned features, but I'm not 100% sure.

File: scrapy_poet/_request_fingerprinter.py
Comment (kmike) [inline scrapy_poet/_request_fingerprinter.py:123]: not a big deal, but the cache is not used in this case

File: scrapy_poet/_request_fingerprinter.py
Comment (kmike) [inline scrapy_poet/_request_fingerprinter.py:130]:
```suggestion
            """Return a JSON object as bytes that represents the page params,
```

File: scrapy_poet/_request_fingerprinter.py
Comment (Gallaecio) [reply scrapy_poet/_request_fingerprinter.py:92]: https://github.com/scrapy/scrapy/pull/6181

File: tests/test_request_fingerprinter.py
Comment (Gallaecio) [reply tests/test_request_fingerprinter.py:70]: https://github.com/scrapinghub/scrapy-poet/pull/169 is merged, so later became now :sweat_smile:

File: scrapy_poet/_request_fingerprinter.py
Comment (Gallaecio) [reply scrapy_poet/_request_fingerprinter.py:44]: [Docs for now it is](https://github.com/scrapinghub/scrapy-poet/pull/172/commits/596431b1c271c13144b1fd45546261b53bba5cb3).

File: tests/test_request_fingerprinter.py
Comment (Gallaecio) [reply tests/test_request_fingerprinter.py:70]: Oh, I had already added a test for annotated dependencies, never mind.

**Review by BurnzZ — APPROVED** [review]

**Review by wRAR — APPROVED** [review]

**Review by kmike — APPROVED** [review]

**Pull request merged**
